File: frontend/src/types/intake.ts
export type IntakeStatus = 'collecting' | 'ready' | 'processing' | 'converting' | 'completed' | 'failed' | 'abandoned';

export interface IntakeDraft {
  title?: string;
  subject?: string;
  grade_level?: string;
  audience?: string;
  duration_minutes?: number;
  scenario?: string;
  language?: string;
  course_task?: string;
  teaching_objectives?: string;
  key_points?: string;
  difficulty_points?: string;
  teaching_method?: string;
  style_requirements?: string;
}

export interface IntakeAssumption {
  field: keyof IntakeDraft | string;
  value: unknown;
  reason: string;
}

export interface IntakeConflict {
  field: keyof IntakeDraft | string;
  severity: 'warning' | 'blocking';
  description: string;
  suggestion: string;
}

export interface IntakeSession {
  id: string;
  status: IntakeStatus;
  current_revision: number;
  draft: IntakeDraft;
  field_sources: Record<string, 'user' | 'material' | 'manual' | 'assumption'>;
  missing_fields: string[];
  assumptions: IntakeAssumption[];
  conflicts: IntakeConflict[];
  course_id?: string | null;
  active_turn_id?: string | null;
  model_config_id?: string | null;
  last_failure?: IntakeTurnFailure | null;
}

export interface IntakeTurnFailure {
  turn_id: string;
  code: string;
  message: string;
  retryable: boolean;
  session_status?: Extract<IntakeStatus, 'collecting' | 'ready' | 'failed'>;
}

export interface IntakeMessage {
  id: string;
  turn_id?: string | null;
  role: 'user' | 'assistant' | 'system';
  content: string;
  created_at?: string;
}

export interface IntakeMaterial {
  id: string;
  original_filename: string;
  mime_type: string;
  size_bytes: number;
  parse_status: 'pending' | 'completed' | 'failed';
  summary: string;
  error_message?: string;
}

export interface IntakeDraftUpdatedEvent {
  revision: number;
  draft: IntakeDraft;
  field_sources: IntakeSession['field_sources'];
  missing_fields: string[];
  assumptions: IntakeAssumption[];
  conflicts: IntakeConflict[];
  ready_to_confirm: boolean;
}

File: backend/app/schemas/intake.py
from typing import Any, Literal

from pydantic import BaseModel, Field


REQUIRED_INTAKE_FIELDS = (
    "title",
    "subject",
    "grade_level",
    "audience",
    "duration_minutes",
    "scenario",
    "course_task",
)


class IntakeCreate(BaseModel):
    model_config_id: str | None = None


class IntakeModelUpdate(BaseModel):
    model_config_id: str


class IntakeDraft(BaseModel):
    title: str | None = None
    subject: str | None = None
    grade_level: str | None = None
    audience: str | None = None
    duration_minutes: int | None = Field(default=None, ge=1, le=300)
    scenario: str | None = None
    language: str | None = None
    course_task: str | None = None
    teaching_objectives: str | None = None
    key_points: str | None = None
    difficulty_points: str | None = None
    teaching_method: str | None = None
    style_requirements: str | None = None


class RequirementAssumption(BaseModel):
    field: str
    value: Any
    reason: str


class RequirementConflict(BaseModel):
    field: str
    severity: Literal["warning", "blocking"] = "warning"
    description: str
    suggestion: str = ""


class RequirementAnalysisResult(BaseModel):
    draft: IntakeDraft
    field_sources: dict[str, Literal["user", "material", "manual", "assumption"]] = Field(default_factory=dict)
    missing_fields: list[str] = Field(default_factory=list)
    assumptions: list[RequirementAssumption] = Field(default_factory=list)
    conflicts: list[RequirementConflict] = Field(default_factory=list)
    next_question: str = ""
    ready_to_confirm: bool = False


class IntakeMessageCreate(BaseModel):
    content: str = Field(min_length=1, max_length=12000)
    expected_revision: int = Field(ge=0)


class IntakeDraftPatch(BaseModel):
    field: str
    value: Any
    expected_revision: int = Field(ge=0)


class IntakeConfirm(BaseModel):
    expected_revision: int = Field(ge=0)
    idempotency_key: str = Field(min_length=8, max_length=120)


class IntakeRead(BaseModel):
    id: str
    status: str
    current_revision: int
    draft: dict[str, Any]
    field_sources: dict[str, Any]
    missing_fields: list[str]
    assumptions: list[dict[str, Any]]
    conflicts: list[dict[str, Any]]
    course_id: str | None
    active_turn_id: str | None = None
    model_config_id: str | None = None
    last_failure: dict[str, Any] | None = None
